compiler/rustc_codegen_llvm/src/asm.rs

```diff
@@ -304,6 +304,7 @@ impl AsmBuilderMethods<'tcx> for Builder<'a, 'll, 'tcx> {
             } else if options.contains(InlineAsmOptions::READONLY) {
                 llvm::Attribute::ReadOnly.apply_callsite(llvm::AttributePlace::Function, result);
             }
+            llvm::Attribute::WillReturn.apply_callsite(llvm::AttributePlace::Function, result);
```

Comment:
Doesn't this imply that `pure` `asm!` cannot infinitely loop (e.g. `asm!("0: jmp 0b")`)? AFAIK `willreturn` implies `mustprogress`.

Reply:
This was already previously assumed (because LLVM assumed `willreturn` everywhere). I can update the `asm!` documentation to reflect this.

Reply:
I'm fine with this landing in either form if we need more time to decide on the behaviour we want here – `asm!` is unstable after all. I'll make a note of this in the `asm!` tracking issue.

Reply:
Made a comment [here](https://github.com/rust-lang/rust/issues/72016#issuecomment-784454061).